freqtrade/exchange/binance_public_data.py

```diff
@@ -0,0 +1,286 @@
+"""
+Fetch daily-archived OHLCV data from https://data.binance.vision/
+"""
+
+import asyncio
+import logging
+import zipfile
+from datetime import date, timedelta
+from io import BytesIO
+from typing import Any
+
+import aiohttp
+import ccxt
+import pandas as pd
+from pandas import DataFrame
+
+from freqtrade.enums import CandleType
+from freqtrade.misc import chunks
+from freqtrade.util.datetime_helpers import dt_from_ts, dt_now
+
+
+logger = logging.getLogger(__name__)
+
+
+class Http404(Exception):
+    def __init__(self, msg, date, url):
+        super().__init__(msg)
+        self.date = date
+        self.url = url
+
+
+class BadHttpStatus(Exception):
+    """Not 200/404"""
+
+    pass
+
+
+async def download_archive_ohlcv(
+    candle_type: CandleType,
+    pair: str,
+    timeframe: str,
+    since_ms: int,
+    until_ms: int | None,
+    markets: dict[str, Any] | None = None,
+    stop_on_404: bool = True,
+) -> DataFrame:
+    """
+    Fetch OHLCV data from https://data.binance.vision
+    The function makes its best effort to download data within the time range
+    [`since_ms`, `until_ms`] -- including `since_ms`, but excluding `until_ms`.
+    If `stop_one_404` is True, this returned DataFrame is guaranteed to start from `since_ms`
+    with no gaps in the data.
+
+    :candle_type: Currently only spot and futures are supported
+    :pair: symbol name in CCXT convention
+    :since_ms: the start timestamp of data, including itself
+    :until_ms: the end timestamp of data, excluding itself
+    :param until_ms: `None` indicates the timestamp of the latest available data
+    :markets: the CCXT markets dict, when it's None, the function will load the markets data
+        from a new `ccxt.binance` instance
+    :param stop_on_404: Stop to download the following data when a 404 returned
+    :return: the date range is between [since_ms, until_ms), return an empty DataFrame if no data
+        available in the time range
+    """
+    try:
+        if candle_type == CandleType.SPOT:
+            asset_type_url_segment = "spot"
+        elif candle_type == CandleType.FUTURES:
+            asset_type_url_segment = "futures/um"
+        else:
+            raise ValueError(f"Unsupported CandleType: {candle_type}")
+
+        if markets:
+            symbol = markets[pair]["id"]
+        else:
+            binance = ccxt.binance()
+            binance.load_markets()
+            symbol = binance.markets[pair]["id"]
```

Comment:
Why would this be necessary? it seems to be a pointless fallback. 
we control all call paths to this method (i don't see this as public api users would call directly) - and markets is guaranteed to be loaded at the point you're reaching here.

if it's not loaded - you'd never get to this point anyway as no pairs would be seen as valid.